File: 01_openmp/GEMV-singleGPU/3-gemv-omp-target-data.cpp
#define N 8192
#include "timer.h"

template<typename T>
void gemv(int n, T alpha, const T* __restrict__ A, const T* __restrict__ V, T* __restrict__ Vout)
{
#pragma omp target teams distribute map(to : A[:n * n], V[:n]) map(from : Vout[:n])
  for (int row = 0; row < n; row++)
  {
    T sum                       = T(0);
    const T* __restrict__ A_row = A + row * n;
#pragma omp parallel for reduction(+ : sum)
    for (int col = 0; col < n; col++)
      sum += A_row[col] * V[col];
    Vout[row] = sum * alpha;
  }
}

template<class T>
T* allocate(size_t n)
{
  T* ptr = new T[n];
  std::fill_n(ptr, n, T(1));
#pragma omp target enter data map(to : ptr[:n])
  return ptr;
}

template<class T>
void deallocate(T* ptr, size_t n)
{
#pragma omp target exit data map(delete : ptr[:n])
  delete[] ptr;
}

int main()
{
  auto* A    = allocate<float>(N * N);
  auto* V    = allocate<float>(N);
  auto* Vout = allocate<float>(N);

  {
    Timer local("GEMV");
    gemv(N, 1.0f, A, V, Vout);
  }

#pragma omp target update from(Vout[:N])
  for (int i = 0; i < N; i++)
    if (Vout[i] != N)
    {
      std::cerr << "Vout[" << i << "] != " << N << ", wrong value is " << Vout[i] << std::endl;
      break;
    }

  deallocate(A, N * N);
  deallocate(V, N);
  deallocate(Vout, N);
}

File: 01_openmp/GEMV-singleGPU/3-gemv-omp-target-data.f90
program gemv_omp

USE OMP_LIB

implicit none
integer,parameter:: N=8192
real(8),allocatable :: A(:),V(:),Vout(:)
real(8) :: alpha=1.0
integer :: ti,tj,tk
integer :: err
integer :: val

allocate(A(1:N*N),stat=err)
if(err/=0) print'(a30,i9,i3)', 'ERROR in allocation for A',err
allocate(V(1:N),stat=err)
if(err/=0) print'(a30,i9,i3)', 'ERROR in allocation for V',err
allocate(Vout(1:N),stat=err)
if(err/=0) print'(a30,i9,i3)', 'ERROR in allocation for Vout',err

A(:) = 1.0
V(:) = 1.0

!$omp target enter data map(to:A,V) map(alloc:Vout)

!!starts here
call system_clock(ti,tk)
call gemv(N,alpha,A,V,Vout)
call system_clock(tj,tk)

print'(a20,3x,f12.4)',"total time: ", dble(tj-ti)/dble(tk)

!$omp target update from(Vout)
do val=1,N
   if (int(Vout(val)) .NE. N) then
        write(*,*) "Value does not match at",val,int(Vout(val))
   end if
end do

!$omp target exit data map(delete:A,V,Vout)
deallocate(A)
if(err/=0) print'(a30,i9,i3)', 'ERROR in deallocation for A',err
deallocate(V)
if(err/=0) print'(a30,i9,i3)', 'ERROR in deallocation for V',err
deallocate(Vout)
if(err/=0) print'(a30,i9,i3)', 'ERROR in deallocation for Vout',err

stop
end

!-------------------------------------------------------
subroutine gemv(nval,alpha,A,V,Vout)

USE OMP_LIB
implicit none

integer:: row,col,A_row
integer:: nval,tid
real(8) :: alpha,sum_val
real(8),intent(in) :: A(1:nval*nval),V(1:nval)
real(8),intent(out):: Vout(1:nval)

!$omp target teams distribute map(to:A,V) map(from:Vout) private(sum_val)
do row=1,nval
   sum_val = 0.0
   A_row =(row-1)*nval
   !$omp parallel do reduction(+:sum_val)
   do col=1,nval
      sum_val = sum_val + A(A_row+col)*V(col)
   end do
   Vout(row) = sum_val * alpha
end do
!$omp end target teams distribute
end subroutine
